src/main/java/org/mule/service/http/impl/service/client/GrizzlyHttpClient.java

```diff
@@ -434,49 +445,70 @@ private HttpResponse sendAndWait(HttpRequest request, HttpRequestOptions options
       auxFuture.whenComplete((response, exception) -> {
         if (response != null) {
           if (shouldFollowRedirect(response, options, enableMuleRedirect)) {
-            handleRedirectAsync(request, response, options, currentRedirects, future);
+            try {
+              handleRedirectAsync(request, response, options, currentRedirects, future);
+            } catch (IOException e) {
+              future.completeExceptionally(e);
+            }
           } else {
             future.complete(response);
           }
         } else {
           future.completeExceptionally(exception);
         }
       });
-
-      asyncHttpClient.executeRequest(createGrizzlyRequest(request, options), asyncHandler);
+      asyncHttpClient.executeRequest(grizzlyRequest, asyncHandler);
     } catch (Exception e) {
       future.completeExceptionally(e);
     }
     return future;
   }
 
   private void handleRedirectAsync(HttpRequest request, HttpResponse response, HttpRequestOptions options,
-                                   int currentRedirects, CompletableFuture<HttpResponse> future) {
-    if (currentRedirects > MAX_REDIRECTS) {
+                                   int currentRedirects, CompletableFuture<HttpResponse> future)
+      throws IOException {
+    if (currentRedirects >= MAX_REDIRECTS) {
       future.completeExceptionally(new MaxRedirectException());
       return;
     }
 
-    HttpRequest redirectRequest = createRedirectRequest(response, request);
-    sendAsync(redirectRequest, options, currentRedirects + 1).whenComplete((redirectResponse, redirectException) -> {
-      if (redirectResponse != null) {
-        future.complete(redirectResponse);
-      } else {
-        future.completeExceptionally(redirectException);
-      }
-    });
+    HttpRequest redirectRequest = createRedirectRequest(response, request, options);
+    Request grizzlyRequest = createGrizzlyRedirectRequest(redirectRequest, response, options);
+    sendAsync(redirectRequest, grizzlyRequest, options, currentRedirects + 1)
+        .whenComplete((redirectResponse, redirectException) -> {
+          if (redirectResponse != null) {
+            future.complete(redirectResponse);
+          } else {
+            future.completeExceptionally(redirectException);
+          }
+        });
+  }
+
+  protected Request createGrizzlyRedirectRequest(HttpRequest request, HttpResponse response, HttpRequestOptions options)
```

Comment:
Indicate what this method does. If the intention is to replace the cookies or if should be override in some use case.

Reply:
Comment added.